plugins/flatpak/gs-flatpak.c

```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+			     "failed to copy flatpak app to removable "
+			     "media: `flatpak create-usb` failed with "
+			     "status %d", status);
+	}
+
+	g_cancellable_disconnect (helper->cancellable, helper->cancelled_id);
+	g_spawn_close_pid (pid);
+
+	helper->finished = TRUE;
+
+	/* once the copy terminates (successfully or not), set plugin status to
+	 * update UI accordingly */
+	gs_plugin_status_update (helper->self->plugin, helper->app,
+				 GS_PLUGIN_STATUS_FINISHED);
+
+	// update metadata to ensure the USB category includes this new app and
+	// we can immediately download an app, copy to USB, delete from
+	// computer, and install from USB without leaving the app page
+	gs_flatpak_refresh (helper->self, 0, GS_PLUGIN_REFRESH_FLAGS_METADATA,
+			    NULL, &error);
+	if (error != NULL) {
+		g_warning ("failed to refresh flatpak metadata after copying "
+			"app to USB: %s", error->message);
+	}
+}
+
+static void
+app_copy_cancelled_cb (GCancellable *cancellable,
+		       gpointer user_data)
+{
+	GPid pid = GPOINTER_TO_INT (user_data);
+
+	// terminate the process which is copying the app
+	kill (pid, SIGTERM);
+}
+
+gboolean
+gs_flatpak_app_copy (GsFlatpak *self,
+		     GsApp *app,
+		     const gchar *dest,
+		     GCancellable *cancellable,
+		     GError **error)
+{
+	// this is used in an async function but we block here until that
+	// returns so we won't auto-free while other threads depend on this
+	g_autoptr (GsFlatpakCopyProcessHelper) helper = NULL;
+	const gchar *app_id;
+	const gchar *app_arch;
+	const gchar *app_branch;
+	g_autofree gchar* app_ref = NULL;
+	gboolean spawn_retval;
+	g_autofree GStrv argv;
+	GPid child_pid;
+	gulong cancelled_id;
+	const gchar* app_collection_id;
+
+	app_id = gs_flatpak_app_get_ref_name (app);
+	app_arch = gs_flatpak_app_get_ref_arch (app);
+	app_branch = gs_flatpak_app_get_ref_branch (app);
+	app_ref = g_strdup_printf ("%s/%s/%s", app_id,
+						     app_arch, app_branch);
+	app_collection_id = gs_flatpak_get_collection_id (self, app);
+	g_debug ("Copying app %s with collection ID: %s", app_ref,
+		 (app_collection_id ? app_collection_id : "(null)"));
+
+	argv = g_malloc (sizeof (gchar*) * 5);
+	argv[0] = "/usr/bin/flatpak";
+	argv[1] = "create-usb";
+	argv[2] = dest;
+	argv[3] = app_ref;
+	argv[4] = NULL;
+
+	gs_plugin_status_update (self->plugin, app, GS_PLUGIN_STATUS_COPYING);
+
+	spawn_retval = g_spawn_async (".",
+				      argv,
+				      NULL,
+				      G_SPAWN_DO_NOT_REAP_CHILD,
+				      NULL,
+				      NULL,
+				      &child_pid,
+				      error);
+
+	if (spawn_retval) {
+		cancelled_id = g_cancellable_connect (cancellable,
+						      G_CALLBACK (app_copy_cancelled_cb),
+						      GINT_TO_POINTER (child_pid),
+						      NULL);
+
+		helper = gs_flatpak_copy_process_helper_new (self, app,
+							     cancellable,
+							     cancelled_id);
+		g_child_watch_add (child_pid, copy_process_watch_cb, helper);
+	} else if (error && *error) {
+		return FALSE;
```

Comment:
I think the return value should be dependent on the passed error arg.

Reply:
Do you mean independent? Because that makes sense.

Fixed locally.